### PR TITLE
thingy91: Adding support for BT over HCI on Thingy:91

### PR DESCRIPTION
**Pull request by crfosse**
This commit adds support in board files and samples for using the
nRF52840 on the Thingy:91 as a BT controller.

Signed-off-by: Carl Richard Steen Fosse <carl.fosse@nordicsemi.no>

### CODE REVIEW

File: boards/arm/thingy91_nrf52840/Kconfig [hunk 1]
```diff
@@ -11,4 +11,11 @@ config BOARD_ENABLE_DCDC
 	select SOC_DCDC_NRF52X
 	default y
 
+config BOARD_THINGY91_NRF52840_RESET
```
[the remark this answers is not in the document]
Comment (crfosse) [reply boards/arm/thingy91_nrf52840/Kconfig:14]: From what I can see it's not actually used no. I can remove it.

File: boards/arm/thingy91_nrf9160/nrf52840_reset.c [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+#define RESET_NODE DT_NODELABEL(nrf52840_reset)
+
+#if DT_NODE_HAS_STATUS(RESET_NODE, okay)
+
+#define RESET_GPIO_CTRL  DT_GPIO_CTLR(RESET_NODE, gpios)
+#define RESET_GPIO_PIN   DT_GPIO_PIN(RESET_NODE, gpios)
+#define RESET_GPIO_FLAGS DT_GPIO_FLAGS(RESET_NODE, gpios)
+
+int bt_hci_transport_setup(const struct device *h4)
```
[the remark this answers is not in the document]
Comment (lemrey) [reply boards/arm/thingy91_nrf9160/nrf52840_reset.c:19]: The nrf52840 must be reset when the nrf9160 is reset to synchronize HCI.

File: boards/arm/thingy91_nrf9160/nrf52840_reset.c [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	if (err) {
+		return err;
+	}
+
+	/* Wait for the nRF52840 peripheral to stop sending data.
+	 *
+	 * It is critical (!) to wait here, so that all bytes
+	 * on the lines are received and drained correctly.
+	 */
+	k_sleep(K_MSEC(10));
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply boards/arm/thingy91_nrf9160/nrf52840_reset.c:50]: Is it a boot banner sending on UART after nrf52 reset ?
If so we can disable it.

File: boards/arm/thingy91_nrf9160/dts/bindings/nordic,thingy91-nrf52840-reset.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2021 Nordic Semiconductor ASA
```
[the remark this answers is not in the document]
Comment (KAGA164) [reply boards/arm/thingy91_nrf9160/dts/bindings/nordic,thingy91-nrf52840-reset.yaml:1]: Maybe we could create some generic bindings to reset Soc in case if we have more Soc than one on single DK ?

File: boards/arm/thingy91_nrf9160/nrf52840_reset.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (crfosse) [reply boards/arm/thingy91_nrf9160/nrf52840_reset.c:50]: Not sure if that's the issue. Someone else can likely comment on this, but my guess is that the hci_lpuart application tries to communicate with the nRF9160 before synchronizing. The data sent during that period must be purged after the nRF52840 is reset, so that the application starts with no garbage on the lines.

File: boards/arm/thingy91_nrf9160/dts/bindings/nordic,thingy91-nrf52840-reset.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (crfosse) [reply boards/arm/thingy91_nrf9160/dts/bindings/nordic,thingy91-nrf52840-reset.yaml:1]: That sounds like a possibility. Do you have any input on that, @lemrey ?

File: boards/arm/thingy91_nrf9160/nrf52840_reset.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (lemrey) [reply boards/arm/thingy91_nrf9160/nrf52840_reset.c:50]: > When the nRF9160 is reset, we need to reset the nRF52840 chip as well to synchronize HCI. We reset the nRF52 using the pin above, and we want to wait for whatever was being sent by the nRF52 to appear on the UART line and drain it, before we let the nRF52 run again to main and re-initialize HCI.

I have explained this already. It's not the boot banner, the nRF52 chip is sending HCI data and it is reset when the nRF91 is reset so that HCI can be re-initialized cleanly. After the nRF52 chip is reset we need to wait for whatever was being sent on the UART to appear on the other side, and "drain" it, otherwise we will fail HCI initialization. If you remove the sleep you will fail HCI initialization when you reset the nRF91.

File: boards/arm/thingy91_nrf9160/dts/bindings/nordic,thingy91-nrf52840-reset.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (crfosse) [reply boards/arm/thingy91_nrf9160/dts/bindings/nordic,thingy91-nrf52840-reset.yaml:1]: Coming back to this: I don't think it will be very productive to put a lot of effort into making this a generic thing. It's a quite narrow use case, and the customers that requires it for their own board will likely just base their board files on Thingy:91 or the nRF9160 DK.

Trying to make this more generic could just lead to confusion. I feel like the current implementation makes the functionality very clear and understandable.

**Review by KAGA164 — COMMENTED** [review]
LGTM. Please add changelog entry

File: samples/bluetooth/hci_lpuart/boards/thingy91_nrf52840.conf [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# Override prj.conf defaults
+
+CONFIG_GPIO=y
+CONFIG_BT_CTLR=y
+CONFIG_MAIN_STACK_SIZE=1024
+CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=512
+CONFIG_BT_WAIT_NOP=y
+CONFIG_BT_MAX_CONN=16
+CONFIG_BT_TINYCRYPT_ECC=n
```
Comment (tejlmand) [inline samples/bluetooth/hci_lpuart/boards/thingy91_nrf52840.conf:9]: Why is this needed ?
Seems to have dependencies to bt stack causing it to be unavailable and thus `n` per default.

File: samples/bluetooth/hci_lpuart/boards/thingy91_nrf52840.conf [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+CONFIG_UART_1_NRF_HW_ASYNC=y
+CONFIG_UART_1_NRF_HW_ASYNC_TIMER=2
+CONFIG_NRF_SW_LPUART=y
+CONFIG_NRF_SW_LPUART_INT_DRIVEN=y
+
+CONFIG_GPIO_AS_PINRESET=y
+
+CONFIG_USE_SEGGER_RTT=y
+CONFIG_RTT_CONSOLE=y
+CONFIG_UART_CONSOLE=n
```
Comment (tejlmand) [inline samples/bluetooth/hci_lpuart/boards/thingy91_nrf52840.conf:24]: Why is this needed ?
Seems to be `n` per default.

File: samples/nrf9160/lte_ble_gateway/boards/thingy91_nrf9160_ns.conf [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+CONFIG_NRF_SW_LPUART_INT_DRIVEN=y
+
+CONFIG_UART_1_ASYNC=y
+CONFIG_UART_1_INTERRUPT_DRIVEN=n
+CONFIG_UART_1_NRF_HW_ASYNC=y
+CONFIG_UART_1_NRF_HW_ASYNC_TIMER=2
+
+CONFIG_USE_SEGGER_RTT=y
+CONFIG_RTT_CONSOLE=y
+CONFIG_UART_CONSOLE=n
```
Comment (tejlmand) [inline samples/nrf9160/lte_ble_gateway/boards/thingy91_nrf9160_ns.conf:11]: Didn't check this one, but based on other observation, could you please verify if this is needed ?

File: boards/arm/thingy91_nrf9160/CMakeLists.txt [hunk 8]
```diff
@@ -18,3 +18,5 @@ if(CONFIG_BOARD_THINGY91_NRF9160_NS)
 	# because it is used in other CMake files.
 	set(PM_STATIC_YML_FILE ${CMAKE_CURRENT_LIST_DIR}/pm_static.yml CACHE INTERNAL "")
 endif()
+
+zephyr_library_sources(nrf52840_reset.c)
```
Comment (tejlmand) [inline boards/arm/thingy91_nrf9160/CMakeLists.txt:22]: Should this file be conditionally included based on node status ?

See: https://github.com/nrfconnect/sdk-zephyr/blob/c932ad1952970987301719d927fc0232116e908f/cmake/extensions.cmake#L2604-L2619

File: samples/bluetooth/hci_lpuart/README.rst [hunk 9]
```diff
@@ -44,6 +44,11 @@ Testing
 
 The methodology to use to test this sample depends on the host application.
 
+Note
```
Comment (tejlmand) [inline samples/bluetooth/hci_lpuart/README.rst:47]: Should this be a tip instead ?

@umapraseeda fyi

> Note - should contain only essential information that must catch the users' attention even if they weren't reading anything else.

File: samples/nrf9160/lte_ble_gateway/README.rst [hunk 10]
```diff
@@ -121,6 +121,10 @@ After programming the main controller with the sample, you can test it as follow
 #. Select the device from your device list on nRF Cloud, and observe that the sensor data is received from the kit.
 #. Observe that the data is updated in nRF Cloud.
 
+Note
```
Comment (tejlmand) [inline samples/nrf9160/lte_ble_gateway/README.rst:124]: Maybe a tip ?

**Review by tejlmand — COMMENTED** [review]
Some minor observations

File: samples/nrf9160/lte_ble_gateway/README.rst (hunk 10 above)
Comment (crfosse) [reply samples/nrf9160/lte_ble_gateway/README.rst:124]: Fixed.

File: boards/arm/thingy91_nrf9160/CMakeLists.txt (hunk 8 above)
Comment (crfosse) [reply boards/arm/thingy91_nrf9160/CMakeLists.txt:22]: Done.

File: samples/nrf9160/lte_ble_gateway/boards/thingy91_nrf9160_ns.conf (hunk 7 above)
Comment (crfosse) [reply samples/nrf9160/lte_ble_gateway/boards/thingy91_nrf9160_ns.conf:11]: Correct! Fixed.

File: samples/bluetooth/hci_lpuart/boards/thingy91_nrf52840.conf (hunk 6 above)
Comment (crfosse) [reply samples/bluetooth/hci_lpuart/boards/thingy91_nrf52840.conf:24]: Inherited from the nRF9160DK configuration, but I've removed it.

Comment (crfosse) [reply samples/bluetooth/hci_lpuart/boards/thingy91_nrf52840.conf:24]: Correct. Fixed.

File: samples/bluetooth/hci_lpuart/boards/thingy91_nrf52840.conf (hunk 5 above)
Comment (crfosse) [reply samples/bluetooth/hci_lpuart/boards/thingy91_nrf52840.conf:9]: Inherited from the nRF9160DK configuration, but I've removed it.

File: samples/nrf9160/lte_ble_gateway/boards/thingy91_nrf9160_ns.conf (hunk 7 above)
Comment (crfosse) [reply samples/nrf9160/lte_ble_gateway/boards/thingy91_nrf9160_ns.conf]: Fixed.

File: boards/arm/thingy91_nrf9160/CMakeLists.txt (hunk 8 above)
Comment (crfosse) [reply boards/arm/thingy91_nrf9160/CMakeLists.txt]: Did not do this the right way.

Comment (crfosse) [reply boards/arm/thingy91_nrf9160/CMakeLists.txt]: Should be OK now!

File: samples/nrf9160/lte_ble_gateway/boards/thingy91_nrf9160_ns.conf (hunk 7 above)
Comment (crfosse) [reply samples/nrf9160/lte_ble_gateway/boards/thingy91_nrf9160_ns.conf:11]: The build fails when I try to remove it. It is enabled by something else, and it isn't prevented as SERIAL is enabled for the HCI communication to work.

**Review by carlescufi — APPROVED** [review]

**Review by KAGA164 — APPROVED** [review]
